### ISSUE TITLE
NVCC & C++14 Constexpr

### ISSUE THREAD

**Issue by ax3l** [mpark/variant#70]
With Nvidia's `nvcc`, tested in versions 10.1.243 and 10.2.89, using `gcc` (v4.9-v8) as host compiler, the following example compiles with `-std=c++11` but not with `-std=c++14`.
```c++
// file: main.cu
#include <mpark/variant.hpp>

int main()
{
    return 0;
}
```

The C++14 error is:
```
mpark/variant.hpp: In function ‘constexpr decltype(auto) mpark::visit(Visitor&&, Vs&& ...)’:
mpark/variant.hpp:2017:96: error: parameter packs not expanded with ‘...’:
     return (detail::all({!vs.valueless_by_exception()...})
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                     ^                                                                                                      
mpark/variant.hpp:2017:96: note:         ‘vs’
```

Naive work-around so far:
```diff
diff --git a/include/mpark/variant.hpp b/include/mpark/variant.hpp
index ef496619b..3a400e546 100644
--- a/include/mpark/variant.hpp
+++ b/include/mpark/variant.hpp
@@ -1998,7 +1998,7 @@ namespace mpark {
     return false;
   }
 
-#ifdef MPARK_CPP14_CONSTEXPR
+#if defined(MPARK_CPP14_CONSTEXPR) && !defined(__NVCC__)
   namespace detail {
 
     inline constexpr bool all(std::initializer_list<bool> bs) {
```

**Comment by ax3l** [comment]
Cross-link Nvidia Bug report: 2767419

**Comment by ax3l** [comment]
The problem is likely in `detail::all` which uses a variadic `using` construct:
https://github.com/mpark/variant/blob/v1.4.0/include/mpark/lib.hpp#L462-L464

**Comment by ax3l** [comment]
Another minimal reproducer developed by @maxpkatz :
```C++
#include <initializer_list>namespace detail {    
    bool f(std::initializer_list<bool> y) { return true; }
}  // namespace detail
template <typename... X>
bool g(X &&... x) {
    // breaks:
    return detail::f({!x.valueless_by_exception()...});
    // works:
    // return detail::f(std::initializer_list<bool>(!x.valueless_by_exception()...));
}

int main() {
    return 0;
}
```

**Comment by mpark** [comment]
Fixed in 3c7fc8266bb46046b42c2dc2663f9f505f0cec28
